### DISCUSSION TITLE
Has anyone been able to setup forwarding to mitmproxy host with pfsense/opnsense (or any BSD-based router)?

### DISCUSSION THREAD

**Discussion by marcosscriven** [mitmproxy/mitmproxy#6063] [Q&A]
I've asked over in the opnsense forum, but noone seems to be able to help https://forum.opnsense.org/index.php?topic=33566.msg162326#msg162326

I see in the docs I have to ensure the destination IP is preserved: https://docs.mitmproxy.org/stable/concepts-modes/#transparent-proxy:

> This distinction is important: when the packet arrives at the mitmproxy machine, it must still be addressed to the target server. This means that Network Address Translation should not be applied before the traffic reaches mitmproxy, since this would remove the target information, leaving mitmproxy unable to determine the real destination

However, I can't seem to work out how to actually do that on my opnsense router. I've tried NAT, but as the mitmproxy docs say, that doesn't work. In Wireshark, I see the traffic hitting, but the destination IP is the mitmproxy host itself, rather than the original server on the internet.

I'm trying to forward anything from `192.168.90.102` that's destined for port 80, to my mitmproxy host on `192.168.90.3`.

As you can see from this Wireshark cap - the destination is the mitmhost, and so the TCP connection is immediately closed:

![Screenshot 2023-04-16 at 08 11 03](https://user-images.githubusercontent.com/752263/232280312-2e803346-1c11-4a6b-bdb3-bafb38f8e1d2.jpg)

**Comment by xkpx64** [comment]
It may help you, it works for me only if i use mitm-regular mode / transparent i can't still run .
No modifications on opnsense itself / or *except install mitmproxy* on  VM

![s](https://github.com/mitmproxy/mitmproxy/assets/5773378/24a793d1-576d-413b-bf75-acf03484cfba)
![s1](https://github.com/mitmproxy/mitmproxy/assets/5773378/52e187e9-73d1-4e09-a30a-ff945dd805a7)
![s3](https://github.com/mitmproxy/mitmproxy/assets/5773378/59d29e87-87c2-48d9-b9cb-d040a3ae3edb)

**Comment by marvinwankersteen** [comment]
The same thing happened to me, and the problem occurs when the host to be monitored is in the same subnet as mitmproxy. If mitmproxy is in a different subnet, the port forwarding rule is sufficient because OPNsense sees the responses, because it is the default gateway, and can rewrite the IPs.

In the same subnet, an outbound rule (SNAT) on this subnet must be created so that the source IP are changed to the interface IP of the OPNsense.

If this rule is missing, mitmproxy initially receives all packets (SYN), but mitmproxy sends the responses (SYN-ACK) directly back to the original source, which then responds with a TCP-RST because it does not expect the response from the mitmproxy IP.
